// ionpad-winston/index.js
/* winston logging initialization */
var winston = require('winston');
var Syslog = require('winston-syslog').Syslog;
var syslog_config = require('winston-syslog').Config;

var MAX_ERRORS = 25;

function configure(name, level, to_console) {

    var times = 0;

    if (winston.config.syslog.levels[level] === undefined) {
        level = 'debug';
    }

    winston.remove(winston.transports.Console);
    if (to_console) {
        winston.add(winston.transports.Console,
                    { level: level });
    }

    var syslog = new Syslog({
        level: level,
        protocol: 'unix-connect',
        path: '/dev/log',
        facility: 'local7',
        app_name: name
    });

    syslog.on('error', function(err) {
        console.error(err);
        ++ times;
        if (times > MAX_ERRORS) {
            winston.remove(syslog);
        } else {
            syslog.socket.close();
            syslog.socket = null;
        }
    });

    syslog.on('logged', function() {
        times = 0;
    });

    winston.add(syslog, null, true);
    winston.setLevels(syslog_config.levels);
}

exports.configure = configure;
